sonar-scanner-engine/src/main/java/org/sonar/scanner/report/ComponentsPublisher.java

```diff
@@ -19,114 +19,147 @@
-    Resource r = batchComponent.resource();
+  private void recursiveWriteComponent(DefaultInputComponent component, ScannerReportWriter writer) {
     ScannerReport.Component.Builder builder = ScannerReport.Component.newBuilder();
 
     // non-null fields
-    builder.setRef(batchComponent.batchId());
-    builder.setType(getType(r));
+    builder.setRef(component.batchId());
+    builder.setType(getType(component));
 
     // Don't set key on directories and files to save space since it can be deduced from path
-    if (batchComponent.isProjectOrModule()) {
+    if (component instanceof InputModule) {
+      DefaultInputModule inputModule = (DefaultInputModule) component;
       // Here we want key without branch
-      ProjectDefinition def = reactor.getProjectDefinition(batchComponent.key());
-      builder.setKey(def.getKey());
-    }
+      builder.setKey(inputModule.key());
 
-    // protocol buffers does not accept null values
+      // protocol buffers does not accept null values
+      String name = getName(inputModule);
+      if (name != null) {
+        builder.setName(name);
+      }
+      String description = getDescription(inputModule);
+      if (description != null) {
+        builder.setDescription(description);
+      }
 
-    if (batchComponent.isFile()) {
-      builder.setIsTest(ResourceUtils.isUnitTestFile(r));
-      builder.setLines(((InputFile) batchComponent.inputComponent()).lines());
-    }
-    String name = getName(r);
-    if (name != null) {
-      builder.setName(name);
+      writeVersion(inputModule, builder);
     }
-    String description = getDescription(r);
-    if (description != null) {
-      builder.setDescription(description);
+
+    if (component.isFile()) {
+      DefaultInputFile file = (DefaultInputFile) component;
+      builder.setIsTest(file.type() == InputFile.Type.TEST);
+      builder.setLines(file.lines());
+
+      String lang = getLanguageKey(file);
+      if (lang != null) {
+        builder.setLanguage(lang);
+      }
     }
-    String path = r.getPath();
+
+    String path = getPath(component);
     if (path != null) {
       builder.setPath(path);
     }
-    String lang = getLanguageKey(r);
-    if (lang != null) {
-      builder.setLanguage(lang);
-    }
-    for (BatchComponent child : batchComponent.children()) {
-      builder.addChildRef(child.batchId());
+
+    Collection<InputComponent> children = componentTree.getChildren(component);
+
+    for (InputComponent child : children) {
+      if (!shouldSkip(child)) {
+        builder.addChildRef(((DefaultInputComponent) child).batchId());
+      }
     }
-    writeLinks(batchComponent, builder);
-    writeVersion(batchComponent, builder);
+    writeLinks(component, builder);
     writer.writeComponent(builder.build());
 
-    for (BatchComponent child : batchComponent.children()) {
-      recursiveWriteComponent(child, writer);
+    for (InputComponent child : children) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Combine this loop with the one that starts on line 106. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3047)